packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -0,0 +1,287 @@
+}
+
+interface ILocalSequencedClient extends ISequencedClient {
+    shouldHaveLeft?: boolean;
+}
+
+export class ConnectionStateHandler {
+    private firstConnection = true;
+    private _manualReconnectInProgress = false;
+    private _messageCountAfterDisconnection: number = 0;
+    private _connectionState = ConnectionState.Disconnected;
+    private pendingClientId: string | undefined;
+    private _clientId: string | undefined;
+    private readonly _connectionTransitionTimes: number[] = [];
+
+    private get protocolHandler(): ProtocolOpHandler {
+        return this.protocolHandlerGetter();
+    }
+
+    private get context() {
+        return this.containerContextGetter();
+    }
+
+    private get audience() {
+        return this.audienceGetter();
+    }
+
+    public get connectionState(): ConnectionState {
+        return this._connectionState;
+    }
+
+    public get connected(): boolean {
+        return this.connectionState === ConnectionState.Connected;
+    }
+
+    public get clientId(): string | undefined {
+        return this._clientId;
+    }
+
+    public get messageCountAfterDisconnection() {
+        return this._messageCountAfterDisconnection;
+    }
+
+    public set messageCountAfterDisconnection(value) {
+        this._messageCountAfterDisconnection = value;
+    }
+
+    public get connectionTransitionTimes() {
+        return this._connectionTransitionTimes;
+    }
+
+    public get manualReconnectInProgress() {
+        return this._manualReconnectInProgress;
+    }
+
+    public set manualReconnectInProgress(value: boolean) {
+        this._manualReconnectInProgress = value;
+    }
+
+    constructor(
+        private readonly container: Container,
+        private readonly deltaManager: DeltaManager,
+        private readonly logger: ITelemetryLogger,
+        private readonly audienceGetter: () => Audience,
+        private readonly containerContextGetter: () => ContainerContext,
+        private readonly protocolHandlerGetter: () => ProtocolOpHandler,
+    ) {
+        this.deltaManager.on(connectEventName, (details: IConnectionDetails, opsBehind?: number) => {
+            const oldState = this._connectionState;
+            this._connectionState = ConnectionState.Connecting;
+
+            // Stash the clientID to detect when transitioning from connecting (socket.io channel open) to connected
+            // (have received the join message for the client ID)
+            // This is especially important in the reconnect case. It's possible there could be outstanding
+            // ops sent by this client, so we should keep the old client id until we see our own client's
+            // join message. after we see the join message for out new connection with our new client id,
+            // we know there can no longer be outstanding ops that we sent with the previous client id.
+            this.pendingClientId = details.clientId;
+
+            container.emit(connectEventName, opsBehind);
+
```

Comment:
as Vald called out, its best not to emit on other classes, could this class just emit, and the container could listen and re-emit?

Reply:
Done